# A beginnerâ€™s guide to low-coverage whole genome sequencing for population genomics

*Runyang Nicolas Lou, Arne Jacobs, Aryn Wilder, Nina O. Therkildsen*

<br> 

This GitHub repository contains our entire simulation and analysis pipeline. Below are links to markdown files containing the pipeline, and the shell scripts and SLiM scripts used in the pipeline are available in the `shell_scripts` and `slim_scripts` folder.

<br> 

## Section 4.1. Population genomic inference for single populations

[Simulation workflow with neutral simulation](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/simulation_workflow_neutral_sim.md)

[Data analysis with neutral simulation and the Samtools GL model (including pool-seq comparison)](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/data_analysis_neutral.md)

[Data analysis with neutral simulation and the GATK GL model (including pool-seq comparison)](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/data_analysis_neutral_gatk.md)

[Simulation workflow with neutral simulation, with uneven input across individuals](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/simulation_workflow_neutral_sim_uneven_input.md)

[Data analysis with neutral simulation, with uneven input across individuals (including pool-seq comparison)](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/data_analysis_neutral_uneven_input.md)

<br> 

## Section 4.2. Inference of spatial structure

[Simulation workflow with spatially structured populations](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/simulation_workflow_spatial_pop_sim.md)

[Data analysis with spatially structured populations](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/data_analysis_spatial_pop.md)

[Simulation workflow with spatially structured populations and lower migration rate](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/simulation_workflow_spatial_pop_sim_lower_migration.md)

[Data analysis with spatially structured populations and lower migration rate](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/data_analysis_spatial_pop_lower_m.md)

[Simulation workflow and data analysis with spatially structured populations and a longer chromosome](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/simulation_workflow_spatial_pop_sim_longer_chr.md)

<br> 

## Section 4.3. Scans for divergent selection in the face of gene flow

[Simulation workflow with divergent selection](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/simulation_workflow_two_pop_sim_fixed_m2_pos.md)

[Data analysis with divergent selection simulation (including RAD-seq comparison)](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/data_analysis_two_pop_fixed_m2_pos.md)

[Data analysis with divergent selection simulation with a smaller population (including RAD-seq comparison)](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/data_analysis_two_pop_lower_s_lower_r.md)

<br> 

## Section 5. Application to empirical data

[Data acquisition and analysis with Heliconius butterflies](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/empirical_data.md)

<br> 

## Box 4. Using imputation to bolster genotype estimation from lcWGS

[Simulation workflow with neutral simulation used for genotype imputation](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/simulation_workflow_neutral_sim_with_replacement.md)

[Imputation without a reference panel](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/imputation.md)

<br> 

## Miscellaneous

[Figures for presentations and the paper](https://github.com/therkildsen-lab/lcwgs-simulation/blob/master/markdowns/figures.md)

<br> 
