---
title: "Data analysis with neutral simulation"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r eval=T, warning=F, message=F}
library(tidyverse)
library(cowplot)
```

## Read in the ancestral states

```{r eval=T, warning=F, message=F}
ancestral <- read_csv("../neutral_sim/rep_1/slim/ancestral.fasta")[[1]] %>%
  str_split(pattern="") %>%
  .[[1]] %>%
  bind_cols(ancestral=., position=1:30000000)
```

## Read mutation and substitution file

```{r eval=T, warning=F, message=F}
## Read in the mutation file outputted by SLiM
mutations <- read_delim("../neutral_sim/rep_1/slim/mutations.txt", delim = " ", col_names = F) %>%
  transmute(type=X6, position=X7+1, base=X13, frequency=X12/2000) %>%
  left_join(ancestral, by="position") %>%
  group_by(type, position, ancestral, base) %>%
  summarise(frequency=sum(frequency)) %>%
  ungroup()
## Read in the substitutions file outputted by SLiM
## This is necessary because mutations can happen again after one fixation, so frequencies from the mutation file do not always reflect the true derived allele frequency
substitutions <- read_delim("../neutral_sim/rep_1/slim/substitutions.txt", delim = " ", skip=2, col_names = F) %>%
  transmute(type=X3, position=X4+1, base=X10, generation=X9) %>%
  group_by(type, position) %>%
  filter(generation==max(generation)) %>%
  ungroup() %>%
  left_join(ancestral, by="position") %>%
  select(-generation) %>%
  filter(base!=ancestral) %>%
  arrange(position)
```

## Data wrangling with the mutation file

The following steps are necessary because there are complications such as back mutations and triallelic loci in the mutation file

```{r eval=T, warning=F, message=F}
## Join mutations and substitutions in a temp table
mutations_final_temp <-  mutations %>%
  spread(key = base, value=frequency) %>%
  full_join(substitutions, by=c("position", "type", "ancestral")) %>%
  arrange(position) %>%
  mutate(base=ifelse(is.na(base), ancestral, base)) %>%
  mutate_all(~replace(., is.na(.), 0)) %>%
  mutate(frequency=1-`A` -`C` -`G` -`T`)
## More wrangling
mutations_final <- mutations_final_temp[1:7] %>%
  gather(key=base, value=frequency, 4:7) %>%
  bind_rows(mutations_final_temp[c(1:3, 8:9)]) %>%
  mutate(frequency=ifelse(base==ancestral, 0, frequency)) %>%
  group_by(type, position, ancestral) %>%
  filter(frequency!=0) %>%
  summarise(frequency=sum(frequency), base=paste0(base, collapse = "")) %>%
  ungroup() %>%
  filter(frequency!=1)
```

## Plot the true site frequency spectrum

```{r eval=T, warning=F, message=F, fig.height=8, fig.width=10}
n_mutations <- dim(mutations_final)[1]
ggplot(mutations_final, aes(x=frequency)) +
  geom_histogram(bins=201) +
  annotate("text", x=0.8, y=150000, label=paste0("n=", n_mutations)) +
  theme_cowplot()
```

## Read maf estimation and join mutation and maf files

```{r eval=F, warning=F, message=F}
i=1
for (coverage in c(0.25,0.5,1,2,4,8)){
  for (sample_size in c(5,10,20,40, 80, 160)){
    ## read in estimated maf
    maf <- read_tsv(paste0("../neutral_sim/rep_1/angsd/bam_list_", sample_size, "_", coverage, "x.mafs.gz")) %>%
      mutate(estimated_frequency=knownEM) %>%
      select(position, major, minor, anc, estimated_frequency, nInd) %>%
      arrange(position)
    ## join estimated maf with true snps and only keep the snps that exist in both data frames
    joined_frequency <- inner_join(mutations_final, maf, by="position") %>%
      select(-ancestral) %>%
      mutate(coverage=coverage, sample_size=sample_size)
    ## find false negatives
    false_negatives <- anti_join(mutations_final, maf, by="position") %>%
      mutate(coverage=coverage, sample_size=sample_size)
    ## find false positives
    false_positives <- anti_join(maf, mutations_final, by="position") %>%
      mutate(coverage=coverage, sample_size=sample_size)
    ## read in estimated sfs
    sfs <- scan(paste0("../neutral_sim/rep_1/angsd/bam_list_", sample_size, "_", coverage, "x.sfs")) %>%
      enframe(name = frequency) %>%
      mutate(frequency=(0:(sample_size*2))/(sample_size*2), coverage=coverage, sample_size=sample_size)
    ## compile the final files for plotting
    if (i==1){
      joined_frequency_final <- joined_frequency
      false_negatives_final <- false_negatives
      false_positives_final <- false_positives
      sfs_final <- sfs
    } else {
      joined_frequency_final <- bind_rows(joined_frequency_final, joined_frequency)
      false_negatives_final <- bind_rows(false_negatives_final, false_negatives)
      false_positives_final <- bind_rows(false_positives_final, false_positives)
      sfs_final <- bind_rows(sfs_final, sfs)
    }
    i=i+1
  }
}
write_tsv(joined_frequency_final, "../neutral_sim/rep_1/angsd/joined_frequency_final.tsv")
write_tsv(false_negatives_final, "../neutral_sim/rep_1/angsd/false_negatives_final.tsv")
write_tsv(false_positives_final, "../neutral_sim/rep_1/angsd/false_positives_final.tsv")
write_tsv(sfs_final, "../neutral_sim/rep_1/angsd/sfs_final.tsv")
```

```{r eval=T, warning=F, message=F}
joined_frequency_final <- read_tsv("../neutral_sim/rep_1/angsd/joined_frequency_final.tsv")
false_negatives_final <- read_tsv("../neutral_sim/rep_1/angsd/false_negatives_final.tsv")
false_positives_final <- read_tsv("../neutral_sim/rep_1/angsd/false_positives_final.tsv")
sfs_final <- read_tsv("../neutral_sim/rep_1/angsd/sfs_final.tsv")
```

## Plot the estimated site frequency spectrum

These are obtained from `realSFS`.

```{r eval=T, warning=F, message=F, fig.height=12, fig.width=15}
sfs_final_sum <- group_by(sfs_final, coverage, sample_size) %>%
  summarise(n=sum(value))

sfs_final %>%
  ggplot(aes(x=frequency, y=value)) +
  geom_point(size=0.5) +
  geom_line() +
  geom_text(data=sfs_final_sum, x=0.8, y=40000, aes(label=paste0("n=",round(n,0)))) +
  facet_grid(coverage~sample_size) +
  theme_cowplot()

sfs_final %>%
  ggplot(aes(x=frequency, y=value)) +
  geom_col(aes(width = 0.5/sample_size)) +
  geom_text(data=sfs_final_sum, x=0.8, y=40000, aes(label=paste0("n=",round(n,0)))) +
  facet_grid(coverage~sample_size) +
  theme_cowplot()
```

## Plot the estimated allele frequency distribution

These are the histogram of estimated allele frequencies

```{r eval=T, warning=F, message=F, fig.height=12, fig.width=15}
linear_model <- group_by(joined_frequency_final, coverage, sample_size) %>%
  summarise(r_squared=paste0("R^2 == ", round(summary(lm(estimated_frequency~frequency))$r.squared,3)),
            n=paste0("n == ",n()))
joined_frequency_final %>%
  ggplot(aes(x=estimated_frequency)) +
  geom_histogram() +
  geom_text(data=linear_model, x=0.8, y=20000, aes(label=n), parse=T) +
  facet_grid(coverage~sample_size, scales ="free_y") +
  theme_cowplot()
```

## Plot estimated allele frequency vs. true allele frequency

```{r eval=T, warning=F, message=F, fig.height=12, fig.width=15}
joined_frequency_final %>%
  ggplot(aes(x=frequency, y=estimated_frequency)) +
  geom_point(alpha=0.1, size=0.1) +
  geom_smooth(method="lm", color="red", size=1, se = F) +
  geom_text(data = linear_model, x = 0.86, y = 0.25, aes(label=r_squared), color = 'red',  parse = TRUE) +
  geom_text(data = linear_model, x = 0.86, y = 0.12, aes(label=n), color = 'red',  parse = TRUE) +
  facet_grid(coverage~sample_size) +
  theme_cowplot()
```

At low coverage and low sample size, the bias in estimation is quite high. This is because of a non-existent nInd filter and a low minDepth filter. Removing these sites will remove some of these noises but will further reduce the amount of data. 

```{r eval=T, warning=F, message=F, fig.height=12, fig.width=15}
joined_frequency_final_nInd_4 <- filter(joined_frequency_final, nInd>=4)

linear_model_nInd_4 <- group_by(joined_frequency_final_nInd_4, coverage, sample_size) %>%
  summarise(r_squared=paste0("R^2 == ", round(summary(lm(estimated_frequency~frequency))$r.squared,3)),
            n=paste0("n == ",n()))

joined_frequency_final_nInd_4 %>%
  ggplot(aes(x=frequency, y=estimated_frequency)) +
  geom_point(alpha=0.1, size=0.2) +
  geom_smooth(method="lm", color="red", size=1, se = F) +
  geom_text(data = linear_model_nInd_4, x = 0.86, y = 0.25, aes(label=r_squared), color = 'red',  parse = TRUE) +
  geom_text(data = linear_model_nInd_4, x = 0.86, y = 0.12, aes(label=n), color = 'red',  parse = TRUE) +
  facet_grid(coverage~sample_size) +
  theme_cowplot()
```

## Plot error vs. true allele frequency

```{r eval=T, warning=F, message=F, fig.height=12, fig.width=15}
joined_frequency_final %>%
  mutate(error=frequency-estimated_frequency) %>%
  ggplot(aes(x=frequency, y=error)) +
  geom_point(alpha=0.1, size=0.2) +
  geom_smooth(method="lm", color="red", size=1, se = F) +
  geom_text(data = linear_model, x = 0.86, y = -0.6, aes(label=n), color = 'red',  parse = TRUE) +
  facet_grid(coverage~sample_size) +
  theme_cowplot()
```

## Check the SNPs with highest error

```{r eval=T, warning=F, message=F}
filter(joined_frequency_final, coverage==8, sample_size==160) %>%
  mutate(error=frequency-estimated_frequency) %>%
  arrange(desc(abs(error))) %>%
  head(n=20)
```

## True frequency distribution of false negatives

```{r eval=T, warning=F, message=F, fig.height=10, fig.width=15}
false_negatives_final_count <- count(false_negatives_final, coverage, sample_size)
ggplot(false_negatives_final, aes(x=frequency)) +
  geom_histogram() +
  geom_text(data=false_negatives_final_count, x=0.3, y=200000, aes(label=paste0("n=", n))) +
  facet_grid(coverage~sample_size) +
  theme_cowplot()
```

## Esimated frequency distribution of false positives

```{r eval=T, warning=F, message=F, fig.height=10, fig.width=15}
false_positives_final_count <- count(false_positives_final, coverage, sample_size)
ggplot(false_positives_final, aes(x=estimated_frequency)) +
  geom_histogram() +
  geom_text(data=false_positives_final_count, x=0.8, y=1500, aes(label=paste0("n=", n))) +
  facet_grid(coverage~sample_size) +
  theme_cowplot()
```
