---
title: "Simulation workflow for neutral simulation"
output: 
  github_document: 
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=F, warning=F}
library(tidyverse)
```

# Neutral simulation

## Create a shell script to run SLiM with nohup

```{r eval=F}
shell_script <- "#!/bin/bash
REP_ID=$1
# Create output directory
if [ ! -d /workdir/lcwgs-simulation/neutral_sim/rep_$REP_ID ]; then
  mkdir /workdir/lcwgs-simulation/neutral_sim/rep_$REP_ID
  cd /workdir/lcwgs-simulation/neutral_sim/rep_$REP_ID
  mkdir angsd
  mkdir bam
  mkdir fasta
  mkdir fastq
  mkdir sample_lists
  mkdir slim
fi
# Run SLiM 
/programs/SLiM-3.3/bin/slim \\
  -d REP_ID=$REP_ID  \\
  -d MUTATION_RATE=100e-8 \\
  -d REC_RATE=250e-8 \\
  -d CHR_LENGTH=30000000 \\
  -d POP_SIZE=1000 \\
  -d SAMPLE_SIZE=2000 \\
  -d \"OUT_PATH='/workdir/lcwgs-simulation/neutral_sim/'\" \\
  /workdir/lcwgs-simulation/slim_scripts/neutral_sim.slim"
write_lines(shell_script, "../shell_scripts/neutral_sim.sh")
```

## Run the shell script for SLiM simulation on server

```{bash eval=F}
for k in 1; do
  nohup bash /workdir/lcwgs-simulation/shell_scripts/neutral_sim.sh $k > '/workdir/lcwgs-simulation/nohups/neutral_sim_'$k'.nohup' &
done
```

## Create a shell script to run ART with nohup

```{r eval=F}
shell_script <-"#!/bin/bash
REP_ID=$1
DIR=${2:-/workdir/lcwgs-simulation/neutral_sim/}

OUT_DIR=$DIR'rep_'$REP_ID'/'
N_CORE_MAX=30
## Generate sam files
COUNT=0
for i in {1..2000}; do
  /workdir/programs/art_bin_MountRainier/art_illumina \\
  -ss HS25 \\
  -sam \\
  -i $OUT_DIR'fasta/derived_'$i'.fasta' \\
  -p \\
  -na \\
  -l 150 \\
  -f 10 \\
  -m 500 \\
  -s 75 \\
  -rs $(($REP_ID*10000+$i)) \\
  -o $OUT_DIR'bam/derived_'$i &
  COUNT=$(( COUNT + 1 ))
  if [ $COUNT == $N_CORE_MAX ]; then
    wait
    COUNT=0
  fi
done
wait
## Generate bam files
COUNT=0
for i in {1..2000}; do
  samtools view -bS -F 4 $OUT_DIR'bam/derived_'$i'.sam' > $OUT_DIR'bam/derived_'$i'.bam' &
  COUNT=$(( COUNT + 1 ))
  if [ $COUNT == $N_CORE_MAX ]; then
    wait
    COUNT=0
  fi
done
wait
## Sort bam files
for i in {1..2000}; do
  rm $OUT_DIR'bam/derived_'$i'.sam'
  mv $OUT_DIR'bam/derived_'$i'1.fq' $OUT_DIR'fastq/'
  mv $OUT_DIR'bam/derived_'$i'2.fq' $OUT_DIR'fastq/'
done"
write_lines(shell_script, "../shell_scripts/art_neutral_sim.sh")
```

## Run the shell script for ART simulation on server

```{bash eval=F}
for k in 1; do
  nohup bash /workdir/lcwgs-simulation/shell_scripts/art_neutral_sim.sh $k > '/workdir/lcwgs-simulation/nohups/art_neutral_sim_'$k'.nohup' &
done
```

## Merge, sort, and subsample bam files

```{r eval=F}
shell_script <-"#!/bin/bash
REP_ID=$1
DIR=${2:-/workdir/lcwgs-simulation/neutral_sim/}

OUT_DIR=$DIR'rep_'$REP_ID'/'
N_CORE_MAX=30
## merge
COUNT=0
for k in {1..1000}; do
  i=$((2*$k-1))
  j=$((2*$k))
  ## merge
  samtools merge $OUT_DIR'bam/sample_'$k'.bam' \\
  $OUT_DIR'bam/derived_'$i'.bam' \\
  $OUT_DIR'bam/derived_'$j'.bam' &
  COUNT=$(( COUNT + 1 ))
  if [ $COUNT == $N_CORE_MAX ]; then
    wait
    COUNT=0
  fi
done
wait
## sort
COUNT=0
for k in {1..1000}; do
  samtools sort -o $OUT_DIR'bam/sample_'$k'_sorted.bam' $OUT_DIR'bam/sample_'$k'.bam' &
  COUNT=$(( COUNT + 1 ))
  if [ $COUNT == $N_CORE_MAX ]; then
    wait
    COUNT=0
  fi
done
wait
## subsample
COUNT=0
for k in {1..1000}; do
  for j in {0.25,0.5,1,2,4,8}; do
    samtools view \\
    -s `awk -v j=$j 'BEGIN { print j / 20 }'` \\
    -b $OUT_DIR'bam/sample_'$k'_sorted.bam' \\
    > $OUT_DIR'bam/sample_'$k'_sorted_'$j'x.bam' &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
for k in {1..1000}; do
  i=$((2*$k-1))
  j=$((2*$k))
  ## delete intermediate files
  #rm $OUT_DIR'bam/derived_'$i'.bam' 
  #rm $OUT_DIR'bam/derived_'$j'.bam' 
  rm $OUT_DIR'bam/sample_'$k'.bam' 
done"
write_lines(shell_script, "../shell_scripts/merge_sort_subsample_neutral_sim.sh")
```

## Run the shell script for merging, sorting, and subsampling

```{bash eval=F}
for k in 1; do
  nohup bash /workdir/lcwgs-simulation/shell_scripts/merge_sort_subsample_neutral_sim.sh $k > '/workdir/lcwgs-simulation/nohups/merge_sort_subsample_neutral_sim_'$k'.nohup' &
done
```

## Make bam lists

```{r eval=F}
for (coverage in c(0.25,0.5,1,2,4,8)){
  for (sample_size in c(5,10,20,40,80,160)){
    for (i in 1:sample_size){
      if (i==1){
        write_lines(paste0("/workdir/lcwgs-simulation/neutral_sim/rep_1/bam/sample_", i, "_sorted_", coverage, "x.bam"), paste0("../neutral_sim/rep_1/sample_lists/bam_list_", sample_size, "_", coverage, "x.txt"))
      } else {
        write_lines(paste0("/workdir/lcwgs-simulation/neutral_sim/rep_1/bam/sample_", i, "_sorted_", coverage, "x.bam"), paste0("../neutral_sim/rep_1/sample_lists/bam_list_", sample_size, "_", coverage, "x.txt"), append = T)
      }
    }
  }
}
```

## Index the ancestral fasta file

```{bash eval=F}
for REP_ID in 1; do
  OUT_DIR='/workdir/lcwgs-simulation/neutral_sim/rep_'$REP_ID'/'
  samtools faidx $OUT_DIR'slim/ancestral.fasta'
done
```

# Run ANGSD using the Samtool genotype likelihood model (`-GL 1`)

## Get shell script for SNP calling

```{r eval=F}
## Note that I used -doMajorMinor 5, using the ancestral sequence to determine major and minor alleles
shell_script <-"#!/bin/bash
BASE_DIR=${1:-/workdir/lcwgs-simulation/neutral_sim/rep_1/}
N_CORE_MAX=28

## SNP calling
COUNT=0
for COVERAGE in {0.25,0.5,1,2,4,8}; do
  for SAMPLE_SIZE in {5,10,20,40,80,160}; do
    /workdir/programs/angsd0.931/angsd/angsd \\
    -b $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \\
    -anc $BASE_DIR'slim/ancestral.fasta' \\
    -out $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x' \\
    -GL 1 -doGlf 2 -doMaf 1 -doMajorMinor 5 \\
    -doCounts 1 -doDepth 1 -dumpCounts 3 \\
    -P 1 -SNP_pval 1e-6 -rmTriallelic 1e-6 \\
    -setMinDepth 2 -minInd 1 -minMaf 0.0005 -minQ 20 &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done"
write_lines(shell_script, "../shell_scripts/snp_calling_neutral_sim.sh")
```

Note: To deal with triallelic loci, I tried `-doMaf 2`, but it turned out that `-doMaf 2` does not bin all alleles other than the ancestral allele. It takes a weighted average approach instead, and makes the result more difficult to process. So we shoud still use`-doMaf 2`, with a `-rmTriallelic 1e-6` filter to remove triallelic loci. 

## Run the shell script for SNP calling

```{bash eval=F}
nohup bash /workdir/lcwgs-simulation/shell_scripts/snp_calling_neutral_sim.sh \
> '/workdir/lcwgs-simulation/nohups/snp_calling_neutral_sim_1.nohup' &
```

## Get shell script for estimating thetas and Tajima's D

```{r eval=F}
shell_script <- "#!/bin/bash
REP_ID=$1
DIR=${2:-/workdir/lcwgs-simulation/neutral_sim/}

BASE_DIR=$DIR'rep_'$REP_ID'/'
N_CORE_MAX=12

## Get saf file
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/angsd \\
      -bam $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \\
      -out $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x' \\
      -doSaf 1 \\
      -anc $BASE_DIR'slim/ancestral.fasta' \\
      -GL 1 \\
      -P 2 \\
      -doCounts 1 \\
      -setMinDepth 2 &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Get SFS from saf (this need to be broken up into two separate runs due to memory limiations)
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/misc/realSFS \\
      $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.saf.idx' \\
      -P 2 \\
      > $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.sfs' &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Estimate theta
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/angsd \\
      -bam $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \\
      -out $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x' \\
      -doThetas 1 \\
      -doSaf 1 \\
      -pest $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.sfs' \\
      -anc $BASE_DIR'slim/ancestral.fasta' \\
      -GL 1 \\
      -P 2 \\
      -doCounts 1 \\
      -setMinDepth 2 &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Print per-SNP theta
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/misc/thetaStat print \\
      $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.idx' \\
      > $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.tsv' &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Do fixed window theta
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/misc/thetaStat do_stat \\
      $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.idx' \\
      -win 10000 -step 10000 \\
      -outnames $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.idx' &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Do per-chromosome average theta
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/misc/thetaStat do_stat \\
      $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.idx' \\
      -outnames $BASE_DIR'angsd/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.average_thetas.idx' &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done"
write_lines(shell_script, "../shell_scripts/estimate_theta_neutral_sim.sh")
```

## Run the shell script to estimate theta and Tajima's D

```{bash eval=F}
nohup bash /workdir/lcwgs-simulation/shell_scripts/estimate_theta_neutral_sim.sh \
  1 \
  > '/workdir/lcwgs-simulation/nohups/estimate_theta_neutral_sim_1.nohup' &
```

# Experimentation with GATK's GL model instead of the Samtool model

Watterson's theta is consistently underestimated, so I will give GATK's GL model a try (`-GL 2`). I will first try this with 40 samples and 1x coverage.

## With minimal filtering

```{bash eval=F}
mkdir /workdir/lcwgs-simulation/neutral_sim/rep_1/angsd_gatk
## Estimate GL and MAF using -GL 2
BASE_DIR=/workdir/lcwgs-simulation/neutral_sim/rep_1/

## Get saf file
for SAMPLE_SIZE in 40; do
  for COVERAGE in 1; do
    nohup /workdir/programs/angsd0.931/angsd/angsd \
      -bam $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \
      -out $BASE_DIR'angsd_gatk/gatk_ml_test' \
      -doSaf 1 \
      -anc $BASE_DIR'slim/ancestral.fasta' \
      -GL 2 \
      -P 16 \
      -doCounts 1 \
      -setMinDepth 2 \
      > /workdir/lcwgs-simulation/nohups/gatk_ml_test_saf.nohup &
  done
done

## Get SFS from saf 
/workdir/programs/angsd0.931/angsd/misc/realSFS \
  $BASE_DIR'angsd_gatk/gatk_ml_test.saf.idx' \
  -P 16 \
  > $BASE_DIR'angsd_gatk/gatk_ml_test.sfs' &

## Estimate theta
for SAMPLE_SIZE in 40; do
  for COVERAGE in 1; do
    nohup /workdir/programs/angsd0.931/angsd/angsd \
      -bam $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \
      -out $BASE_DIR'angsd_gatk/gatk_ml_test' \
      -doThetas 1 \
      -doSaf 1 \
      -pest $BASE_DIR'angsd_gatk/gatk_ml_test.sfs' \
      -anc $BASE_DIR'slim/ancestral.fasta' \
      -GL 2 \
      -P 16 \
      -doCounts 1 \
      -setMinDepth 2 \
      > /workdir/lcwgs-simulation/nohups/gatk_ml_test_theta.nohup &
  done
done

## Do per-chromosome average theta
/workdir/programs/angsd0.931/angsd/misc/thetaStat do_stat \
  $BASE_DIR'angsd_gatk/gatk_ml_test.thetas.idx' \
  -outnames $BASE_DIR'angsd_gatk/gatk_ml_test.average_thetas.idx' &
```

## With `-setMinDepth 40`

```{bash eval=F}
BASE_DIR=/workdir/lcwgs-simulation/neutral_sim/rep_1/
for SAMPLE_SIZE in 40; do
  for COVERAGE in 1; do
    nohup /workdir/programs/angsd0.931/angsd/angsd \
      -bam $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \
      -out $BASE_DIR'angsd_gatk/gatk_ml_test_mindepth_40' \
      -doSaf 1 \
      -anc $BASE_DIR'slim/ancestral.fasta' \
      -GL 2 \
      -P 16 \
      -doCounts 1 \
      -setMinDepth 40 \
      > /workdir/lcwgs-simulation/nohups/test_saf.nohup &
  done
done

/workdir/programs/angsd0.931/angsd/misc/realSFS \
  $BASE_DIR'angsd_gatk/gatk_ml_test_mindepth_40.saf.idx' \
  -P 16 \
  -tole 1e-07 \
  > $BASE_DIR'angsd_gatk/gatk_ml_test_mindepth_40.sfs' &

for SAMPLE_SIZE in 40; do
  for COVERAGE in 1; do
    nohup /workdir/programs/angsd0.931/angsd/angsd \
      -bam $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \
      -out $BASE_DIR'angsd_gatk/gatk_ml_test_mindepth_40' \
      -doThetas 1 \
      -doSaf 1 \
      -pest $BASE_DIR'angsd_gatk/gatk_ml_test_mindepth_40.sfs' \
      -anc $BASE_DIR'slim/ancestral.fasta' \
      -GL 2 \
      -P 16 \
      -doCounts 1 \
      -setMinDepth 40 \
      > /workdir/lcwgs-simulation/nohups/gatk_ml_test_theta_mindepth_40.nohup &
  done
done

/workdir/programs/angsd0.931/angsd/misc/thetaStat do_stat \
  $BASE_DIR'angsd_gatk/gatk_ml_test_mindepth_40.thetas.idx' \
  -outnames $BASE_DIR'angsd_gatk/gatk_ml_test_mindepth_40.average_thetas.idx' &
```

## Result comparison

```{r message=F, warning=F, fig.height=8, fig.width=10}
## SFS comparison
gatk_sfs <- read_lines("../neutral_sim/rep_1/angsd_gatk/gatk_ml_test.sfs") %>%
  str_split(pattern = " ") %>%
  .[[1]] %>%
  .[1:81] %>%
  as.numeric() %>%
  bind_cols(frequency=0:80/80, count=., method=rep("GATK (MinDepth=2)",81))
gatk_sfs_mindepth_40 <- read_lines("../neutral_sim/rep_1/angsd_gatk/gatk_ml_test_mindepth_40.sfs") %>%
  str_split(pattern = " ") %>%
  .[[1]] %>%
  .[1:81] %>%
  as.numeric() %>%
  bind_cols(frequency=0:80/80, count=., method=rep("GATK (MinDepth=40)",81))
samtools_sfs <- read_lines("../neutral_sim/rep_1/angsd/bam_list_40_1x.sfs") %>%
  str_split(pattern = " ") %>%
  .[[1]] %>%
  .[1:81] %>%
  as.numeric() %>%
  bind_cols(frequency=0:80/80, count=., method=rep("Samtools (MinDepth=2)",81))
bind_rows(gatk_sfs, gatk_sfs_mindepth_40, samtools_sfs) %>%
  filter(frequency>0, frequency<1) %>%
  ggplot() +
  geom_col(aes(x=frequency, y=count)) +
  facet_grid(method~.) +
  cowplot::theme_cowplot()

## chromosome-average theta comparison
gatk_theta <- read_tsv("../neutral_sim/rep_1/angsd_gatk/gatk_ml_test.average_thetas.idx.pestPG") %>%
  transmute(t_w = tW/nSites, t_p = tP/nSites, method = "GATK (MinDepth=2)")
gatk_theta_mindepth_40 <- read_tsv("../neutral_sim/rep_1/angsd_gatk/gatk_ml_test_mindepth_40.average_thetas.idx.pestPG") %>%
  transmute(t_w = tW/nSites, t_p = tP/nSites, method = "GATK (MinDepth=40)")
samtools_theta <- read_tsv("../neutral_sim/rep_1/angsd/bam_list_40_1x.average_thetas.idx.pestPG") %>%
  transmute(t_w = tW/nSites, t_p = tP/nSites, method = "Samtools (MinDepth=2)")
bind_rows(gatk_theta, gatk_theta_mindepth_40, samtools_theta) %>%
  knitr::kable()
```

The GATK model does a better job. The depth filtering also makes a big difference. I will therefore run the entire process (starting from SNP calling) using the GATK model below. For the SFS part, I will use a more stringent depth filter.

# Run ANGSD using the GATK genotype likelihood model (`-GL 2`)

## Get shell script for SNP calling

```{r eval=F}
## Note that I used -doMajorMinor 5, using the ancestral sequence to determine major and minor alleles
shell_script <-"#!/bin/bash
BASE_DIR=${1:-/workdir/lcwgs-simulation/neutral_sim/rep_1/}
N_CORE_MAX=28

## SNP calling
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/angsd \\
    -b $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \\
    -anc $BASE_DIR'slim/ancestral.fasta' \\
    -out $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x' \\
    -GL 2 -doGlf 2 -doMaf 1 -doMajorMinor 5 \\
    -doCounts 1 -doDepth 1 -dumpCounts 3 \\
    -P 1 -SNP_pval 1e-6 -rmTriallelic 1e-6 \\
    -setMinDepth 2 -minInd 1 -minMaf 0.0005 -minQ 20 &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done"
write_lines(shell_script, "../shell_scripts/snp_calling_gatk_neutral_sim.sh")
```

Note: To deal with triallelic loci, I tried `-doMaf 2`, but it turned out that `-doMaf 2` does not bin all alleles other than the ancestral allele. It takes a weighted average approach instead, and makes the result more difficult to process. So we shoud still use`-doMaf 2`, with a `-rmTriallelic 1e-6` filter to remove triallelic loci. 

## Run the shell script for SNP calling

```{bash eval=F}
nohup bash /workdir/lcwgs-simulation/shell_scripts/snp_calling_gatk_neutral_sim.sh \
> '/workdir/lcwgs-simulation/nohups/snp_calling_gatk_neutral_sim_1.nohup' &
```

## Get shell script for estimating thetas and Tajima's D

A more stringent mininum depth filter, which equals to `sample_size * coverage`, is used in this step.

```{r eval=F}
shell_script <- "#!/bin/bash
REP_ID=$1
DIR=${2:-/workdir/lcwgs-simulation/neutral_sim/}

BASE_DIR=$DIR'rep_'$REP_ID'/'
N_CORE_MAX=12

## Get saf file
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/angsd \\
      -bam $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \\
      -out $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x' \\
      -doSaf 1 \\
      -anc $BASE_DIR'slim/ancestral.fasta' \\
      -GL 2 \\
      -P 2 \\
      -doCounts 1 \\
      -setMinDepth `awk \"BEGIN {print $SAMPLE_SIZE*$COVERAGE}\"` &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Get SFS from saf (this need to be broken up into two separate runs due to memory limiations)
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/misc/realSFS \\
      $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.saf.idx' \\
      -P 2 \\
      > $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.sfs' &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Estimate theta
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/angsd \\
      -bam $BASE_DIR'sample_lists/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.txt' \\
      -out $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x' \\
      -doThetas 1 \\
      -doSaf 1 \\
      -pest $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.sfs' \\
      -anc $BASE_DIR'slim/ancestral.fasta' \\
      -GL 2 \\
      -P 2 \\
      -doCounts 1 \\
      -setMinDepth `awk \"BEGIN {print $SAMPLE_SIZE*$COVERAGE}\"` &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Print per-SNP theta
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/misc/thetaStat print \\
      $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.idx' \\
      > $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.tsv' &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Do fixed window theta
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/misc/thetaStat do_stat \\
      $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.idx' \\
      -win 10000 -step 10000 \\
      -outnames $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.idx' &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done
wait
## Do per-chromosome average theta
COUNT=0
for SAMPLE_SIZE in {5,10,20,40,80,160}; do
  for COVERAGE in {0.25,0.5,1,2,4,8}; do
    /workdir/programs/angsd0.931/angsd/misc/thetaStat do_stat \\
      $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.thetas.idx' \\
      -outnames $BASE_DIR'angsd_gatk/bam_list_'$SAMPLE_SIZE'_'$COVERAGE'x.average_thetas.idx' &
    COUNT=$(( COUNT + 1 ))
    if [ $COUNT == $N_CORE_MAX ]; then
      wait
      COUNT=0
    fi
  done
done"
write_lines(shell_script, "../shell_scripts/estimate_theta_gatk_neutral_sim.sh")
```

## Run the shell script to estimate theta and Tajima's D

```{bash eval=F}
nohup bash /workdir/lcwgs-simulation/shell_scripts/estimate_theta_gatk_neutral_sim.sh \
  1 \
  > '/workdir/lcwgs-simulation/nohups/estimate_theta_gatk_neutral_sim_1.nohup' &
```